modeshape-jcr/src/main/java/org/modeshape/jcr/RepositoryQueryManager.java

```diff
@@ -249,16 +256,88 @@ protected final QueryEngine queryEngine() {
         }
         return queryEngine;
     }
+    
+    protected void reindex() {
+        RepositoryConfiguration.Reindexing reindexingCfg = repoConfig.getReindexing();
+        boolean async = reindexingCfg.isAsync();
+        RepositoryConfiguration.ReindexingMode mode = reindexingCfg.mode();
+        switch (mode) {
+            case INCREMENTAL: {
+                final ChangeJournal journal = runningState.journal();
+                if (journal == null) {
+                    logger.warn(JcrI18n.warnIncrementalIndexingJournalNotEnabled, repoConfig.getName());
+                    break;
+                } else if (!journal.started()) {
+                    logger.warn(JcrI18n.warnIncrementalIndexingJournalNotStarted, repoConfig.getName());
+                    break;
+                } else {
+                    if (!async) {
+                        reindexIncrementally(journal);
+                    } else {
+                        indexingExecutorService.submit(new Callable<Void>() {
+                            @Override
+                            public Void call() throws Exception {
+                                reindexIncrementally(journal);
+                                return null;
+                            }
+                        }); 
+                    }
+                    break;
+                }
+            }
+            case FULL: {
+                reindexIfNeeded(async);
+                break;
+            }
+            default: {
+                throw new IllegalArgumentException("Unknown indexing mode: " + mode.toString());
+            }
+        }
+    }
+
+    private void reindexIncrementally( ChangeJournal journal ) {
+        // each provider may have a different timestamp when it last updated its indexes successfully. 
+        // so for simplicity we'll use the lowest (earliest) timestamp of them all when reindexing
+        long earliestTimestamp = Long.MAX_VALUE;
+        List<IndexProvider> incrementalIndexingProviders = new ArrayList<>();
+        for (IndexProvider provider : indexManager.getProviders()) {
+            Long latestIndexUpdateTime = provider.getLatestIndexUpdateTime();
+            if (latestIndexUpdateTime == null) {
+                logger.warn(JcrI18n.warnIncrementalIndexingNotSupported, provider.getName());
+                continue;
+            }
+            incrementalIndexingProviders.add(provider);
+            earliestTimestamp = Math.min(earliestTimestamp, latestIndexUpdateTime);
+        }
+        if (incrementalIndexingProviders.isEmpty()) {
+            // none of the providers support incremental reindexing so nothing to do...
+            return;
+        }
+        assert earliestTimestamp != Long.MAX_VALUE;
+        Set<NodeKey> changedNodes = journal.changedNodesSince(earliestTimestamp);
```

Comment:
Isn't it pretty likely that the set of changed node keys might be very large, especially when adding a new process to the cluster? And might that cause memory issues?

I know this is trying to save the amount of work that has to be done by computing the set of changed node keys, but wouldn't it also be possible for the journal to efficiently return an `Iterator<NodeKey>` and then for the `reindexSince(...)` method to use the iterator? The journal could back this by a `Set`, in which case the current behavior is achieved, but it could also back it by an iterator that literally walks the journal and possibly returns a node more than once. The former is much easier to start out with, but the latter is much more scalable for situations where a lot of nodes (perhaps all of them) have been changed. Changing the signature would give us flexibility, while the `reindex` method used on line 326 wouldn't really care.